src/detect-transform-domain.c

```diff
@@ -0,0 +1,274 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Eric Leblond <el@stamus-networks.com>
+ *
+ * Implements the domain extraction transformation
+ */
+
+#include "suricata-common.h"
+
+#include "detect.h"
+#include "detect-engine.h"
+#include "detect-engine-prefilter.h"
+#include "detect-parse.h"
+#include "detect-transform-domain.h"
+#include "detect-engine-build.h"
+
+#include "util-unittest.h"
+#include "util-print.h"
+#include "util-memrchr.h"
+#include "util-memcpy.h"
+#include "rust.h"
+
+static int DetectTransformDomainSetup(DetectEngineCtx *, Signature *, const char *);
+static int DetectTransformTLDSetup(DetectEngineCtx *, Signature *, const char *);
+#ifdef UNITTESTS
+static void DetectTransformDomainRegisterTests(void);
+static void DetectTransformTLDRegisterTests(void);
+#endif
+static void TransformDomain(InspectionBuffer *buffer, void *options);
+static void TransformTLD(InspectionBuffer *buffer, void *options);
+
+void DetectTransformDomainRegister(void)
+{
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].name = "domain";
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].desc = "modify buffer to extract the domain";
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].url = "/rules/transforms.html#domain";
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].Transform = TransformDomain;
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].Setup = DetectTransformDomainSetup;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].RegisterTests = DetectTransformDomainRegisterTests;
+#endif
+    sigmatch_table[DETECT_TRANSFORM_DOMAIN].flags |= SIGMATCH_NOOPT;
+
+    sigmatch_table[DETECT_TRANSFORM_TLD].name = "tld";
+    sigmatch_table[DETECT_TRANSFORM_TLD].desc = "modify buffer to extract the tld";
+    sigmatch_table[DETECT_TRANSFORM_TLD].url = "/rules/transforms.html#tld";
+    sigmatch_table[DETECT_TRANSFORM_TLD].Transform = TransformTLD;
+    sigmatch_table[DETECT_TRANSFORM_TLD].Setup = DetectTransformTLDSetup;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_TRANSFORM_TLD].RegisterTests = DetectTransformTLDRegisterTests;
+#endif
+    sigmatch_table[DETECT_TRANSFORM_TLD].flags |= SIGMATCH_NOOPT;
+}
+
+/**
+ *  \internal
+ *  \brief Extract the dotprefix, if any, the last pattern match, either content or uricontent
+ *  \param det_ctx detection engine ctx
+ *  \param s signature
+ *  \param nullstr should be null
+ *  \retval 0 ok
+ *  \retval -1 failure
+ */
+static int DetectTransformDomainSetup(DetectEngineCtx *de_ctx, Signature *s, const char *nullstr)
+{
+    SCEnter();
+    int r = DetectSignatureAddTransform(s, DETECT_TRANSFORM_DOMAIN, NULL);
+    SCReturnInt(r);
+}
+
+/**
+ * \brief Return the domain, if any, in the last pattern match.
+ *
+ */
+static void TransformDomain(InspectionBuffer *buffer, void *options)
+{
+    const size_t input_len = buffer->inspect_len;
+    uint64_t output_len = 0;
+
+    if (input_len) {
+        uint8_t output[input_len];
```

Comment:
we should avoid stack allocations like this, as they are virtually unbounded

In this case we can do an inplace change?

cc @catenacyber